KOGITO-5208: Lienzo - Migration to native interfaces

romartin opened the pull request:
This PR contains the lienzo migration to J2CL. 

It basically incorporates 3 new modules inside this repo (including all git history being migrated as well):
- lienzo-core
- lienzo-tests
- lienzo-webapp

Also it contains all editors updates in order to make this migration run properly: BPMN, DMN and SceSim

**VSIX for testing editors**: BPMN, DMN and SceSim - uploaded [HERE](https://drive.google.com/file/d/1dTn4ZxzDE9PfPCQ4ypHZhfCeFohXP4xY/view?usp=sharing).

See https://issues.redhat.com/browse/KOGITO-5208

@LuboTerifaj You're the main QE tester for the ticket, so added you as a reviewer
@hasys @Josephblt Added you as reviewers for the BPMN side
@karreiro Added you as reviewer for the DMN side
@yesamer Added you as reviewer for the SceSim side
@tiagobento Added you as reviewer for the framework side


CC Thanks a lot @handreyrc for collaborating in this work, let's be responsive on this PR and try to get this merged ASAP.

Thanks guys!

tiagobento commented:
@romartin Nice work! From what we've already discussed, this PR is looking good! I'll try to take a second look at the files once again later this week. Also, the FDB build broke due to an instability on one of our tests. This is already being addressed on https://github.com/kiegroup/kogito-tooling/pull/535

tiagobento commented:
⚠️ IMPORTANT ⚠️

Don't squash this PR, otherwise the commit history will be lost. Use the `Merge pull request` option.

cc @romartin @karreiro

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/ImageElementProxy.java:
```diff
@@ -0,0 +1,75 @@
+package com.ait.lienzo.client.core.image;
```
Missing license header

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/ImageLoader.java:
```diff
@@ -0,0 +1,188 @@
+
+        RootPanel.get().add(image);
+
+        // @FIXME I removed "isValidDataURL(url) && isValidSVG(url)" as it seems with Elemental2 it all end up at .src. But we should double check this (mdp)
```
Unfinished FIXME

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/ImageLoader.java:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    public static void setVisible(HTMLElement image, boolean visible) {
+        image.style.display = visible ? "" : Style.Display.NONE.getCssName(); // @FIXME check that it should really be "", I just followed GWT UIObject setVisible.
```
Unfinished FIXME

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/ImageLoader.java:
```diff
@@ -0,0 +1,188 @@
+            return null;
+        };
+
+        // @FIXME double check this works, ImageReources can includ clippping information, so I had to case from Elemental2 to GWT (mdp)
```
Unfinished FIXME

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/ImageLoader.java:
```diff
@@ -0,0 +1,188 @@
+            {
+                image.onload = null;
+                image.onerror = null;
+                // @FIXME removed 'naturalHeight' in image I think this is supported by all browser now. Needs double check (mdp)
```
Unfinished FIXME

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/ImageLoader.java:
```diff
@@ -0,0 +1,188 @@
+                // @FIXME removed 'naturalHeight' in image I think this is supported by all browser now. Needs double check (mdp)
+                if (image.naturalHeight + image.naturalWidth == 0 || image.height + image.width == 0) {
+                    // it failed, so undo
+                    // @FIXME check this cast works (mdp)
```
Unfinished FIXME

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/ImageLoader.java:
```diff
@@ -0,0 +1,188 @@
+                image.onload = null;
+                image.onerror = null;
+                // it failed, so undo
+                // @FIXME check this cast works (mdp)
```
Unfinished FIXME

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/AbstractConvolveImageDataFilter.java:
```diff
@@ -0,0 +1,136 @@
+        super(type, node, ctx);
+    }
+
+//    public final void setMatrix(final double... matrix)
```
Commented code, can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/AbstractImageDataFilter.java:
```diff
@@ -0,0 +1,155 @@
+            return;
+        }
+
+        // @FIXME serialization (mdp)
```
Is it still valid? Can we remove it here and below?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/AverageGrayScaleImageDataFilter.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length) {
+//        int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code, can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/BrightnessImageDataFilter.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length, double value) {
+//        int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code, can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/ColorDeltaAlphaImageDataFilter.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length, int r, int g, int b, double v) {
+//        int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/ColorLuminosityImageDataFilter.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length, int r, int g, int b) {
+//        int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/EdgeDetectImageDataFilter.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, Uint8ClampedArray buffArray, int w, int h) {
+//        int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/EmbossImageDataFilter.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length, int width, ImageDataFilterCommonOps fops) {
+//        int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/HueImageDataFilter.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int w, int h, double value, ImageDataFilterCommonOps fops) {
+        //int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/ImageDataFilterCommonOps.java:
```diff
@@ -0,0 +1,369 @@
+
+    // MDP set correct parameters....
+    public void filterTable(Uint8ClampedArray dataArray, FilterTableArray tableArray, int w, int h) {
+        //int[] t1 = Uint8ClampedArray.ConstructorLengthUnionType.of(data).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/ImageDataFilterCommonOps.java:
```diff
@@ -0,0 +1,369 @@
+
+    public void filterLuminosity(Uint8ClampedArray dataArray, int length) {
+
+        int[] data = Js.uncheckedCast(dataArray); //Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/ImageDataFilterCommonOps.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    public int[] getPixel(Uint8ClampedArray dataArray, double x, double y, double w, double h) {
+        //int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/ImageDataFilterCommonOps.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    public void filterTransform(Uint8ClampedArray dataArray, Uint8ClampedArray buffArray, FilterTransformFunction transformer, int w, int h) {
+//        int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/ImageDataFilterCommonOps.java:
```diff
@@ -0,0 +1,369 @@
+    ;
+
+    public boolean hasAlphaChannel(Uint8ClampedArray dataArray, int length) {
+        //int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/InvertColorImageDataFilter.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length) {
+        //int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/LightnessGrayScaleImageDataFilter.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length) {
+        //int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/RGBIgnoreAlphaImageDataFilter.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int length, int r, int g, int b) {
+//        Uint8ClampedArray.ConstructorLengthUnionType temp = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray);
```
Commented code can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/filter/StackBlurImageDataFilter.java:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    private final void filter_(Uint8ClampedArray dataArray, int width, int height, int radius, ImageDataFilterCommonOps fops) {
+        //int[] data = Uint8ClampedArray.ConstructorLengthUnionType.of(dataArray).asIntArray();
```
Commented code, can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/shape/AbstractMultiPathPartShape.java:
```diff
@@ -0,0 +1,1418 @@
+        }
+    }
+
+//    public final Double getMinWidth() {
```
Commented code, can we remove it?

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/shape/AbstractMultiPathPartShape.java:
```diff
@@ -0,0 +1,1418 @@
+                Point2D dxy = new Point2D(0 - xShift, 0 - yShift);
+                shiftPoints(dxy);
+
+                // TODO (mdp) this is a hack for now, as pure lienzo shapes cannot access Wires yet, and no easy way to inject a callable interface.
```
// Unfinished TODO

romartin replied:
done

hasys commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/shape/AbstractPathClipper.java:
```diff
@@ -0,0 +1,263 @@
+import jsinterop.annotations.JsProperty;
+import jsinterop.annotations.JsType;
+
+// @TODO This class and it's children need more checking (mdp)
```
Not finished TODO

romartin replied:
done

hasys reviewed (commented):
Hi @romartin, great job, I like to see removed dependencies and native code. Overall looks like a great improvement to me thank you! 

there are my 2c mostly about missing license header and FIXME/TODO/Uncommented code. Looks like some of TODO/FIXME inherited from lienzo-core, but some of them looks fresh to me, so just to double check.

Also I think we can remove commented code even if it is inherited from lienzo-core. It wasn't changed for years and I think it won't be changed soon. WDYT?

This is the first part of the review, will finish the review later today/tomorrow. Thank you!

hasys reviewed (commented):
Hi @romartin, @handreyrc,

great job, thank you! I cherry-picked my changes to this branch and all works smooth so far. 

Since we are doing several things in this PR already I suggest also remove all dead code (commented code and modules which are not relevant any more).

If I correct we have all lienzo-core uses in this repo now so we can easily remove and check that code doesn't break compilation for any of projects.

I putted quite a lot of comments but most of the missing license headers and commented code, I think it should be fast to address.

However some TODOs and FIXME looks like fresh and related to the migration to the native events so I think it will be good to address as well.

I can help you with fixing issue and moving some commented code to jira issues for better visibility if you need a help to speed up things.

Thank you guys for your hard work!

romartin commented:
Hey @hasys 

Thanks for the early review on code. Yes, I'm am cleaning up those TODOS, or moving to separate tasks if necessary, don't really pay too much attention to commented stuff yet please, on the meanwhile, this PR at least allows testing the stuff, also in your MAC :) So please any feedback is welcome!

Thx!

romartin commented:
Hey @hasys @karreiro @yesamer @LuboTerifaj @Josephblt 

Uploaded the VSIX file if you want to use it for testing this work. It contains BPMN, DMN and SceSim. Link is on the description https://github.com/kiegroup/kogito-editors-java/pull/63#issue-669559318

PS: I'm still working on code cleanups but everything is ready for testing the UI.

Thanks!

karreiro reviewed (commented):
Thank you for this PR, @romartin! 

I've noticed that some feature in the relationship between Decision Services in inner nodes seems lost (as the Gif shows, when I move the DS and the decision node keeps the same position).

Also, while the VSIX seems fine, when I build the PR locally and run `kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-testing`, the toolboxes do not appear anymore, and the text in the node is not centralized.

For testing things locally, I built the kogito-editors-java before running the `-dmn-webapp-kogito-testing` web app. Am I missing something? 

Thanks again for the PR!

romartin commented:
@karreiro thanks for the early feedback! So about the 2 things you mention:

1- It looks an issue, yes! Once adding a node in the DS, it should be a child, so move in conjuntion with the parent. I'll take a look right now and let you know. 

2.- If the toolbox does not appear, it means the jsinterop exports are not enabled, we faced this issue too. So we just updated the README with this flag to be set. The maven build already does that, but if you create a run config in your IDE, you need to set this flag manually. See the flag `-generateJsInteropExports` -> https://github.com/romartin/kogito-editors-java/tree/KOGITO-5208/kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime#running-in-intellij 

So I'll check 1), and plz lemme know if 2) gets fixed once setting the flag.

Thanks! :+1:

romartin commented:
@karreiro Oh just noticed the `generateJsInteropExports` flag was not set for both testing apps for DMN and SceSim, thanks for the feedback! Just added a new commit that fixes it.  :+1:

romartin commented:
@karreiro another update on your comments above https://github.com/kiegroup/kogito-editors-java/pull/63#pullrequestreview-686490904:

1.- Good catch about the containment bug, it also happens in BPMN. It happens once moving a child, which is already inside some parent, somewhere inside same parent. At some point, it "looses" the parent relationship. So working on it, will let you know. Thx for the catch.

2.- Finally added a commit that adds the generateJsInteropExports flag to all GWT apps now, hopefully :) Anyway please remember to set this flag in your IDE run configurations, if any.

Please try to give a try to some other features as well, and provide early feedback, really appreciated! I am focused on trying to close this PR asap, so others can work on top.

Let's keep in touch! 
Thx!

romartin commented:
@karreiro Fixed issue 1) as well - new commit added :) Thanks! 

@karreiro @LuboTerifaj @yesamer @hasys @Josephblt Updated the VSIX file with latest fixes, ready to test (download from the description link https://github.com/kiegroup/kogito-editors-java/pull/63#issue-669559318)

Thanks!

jomarko reviewed (changes requested):
I found an issue, that user an not invoke a context menu, if the expression grid editor is scrolled.

![Screenshot from 2021-06-21 08-05-26](https://user-images.githubusercontent.com/8044780/122714487-a9f80400-d267-11eb-968c-ce2d105e4136.png)

The used model in the ^ is attached into https://issues.redhat.com/browse/KOGITO-5208 as **fee.dmn**. There if you open Decision-2 node, and scroll down to the 7th context entry row, that is defined as List, you will be not able to invoke context menu for adding rows.

jomarko commented:
I found out, that https://github.com/kiegroup/kogito-editors-java/pull/63#pullrequestreview-688057234 is also on master, so I reported a separate issue:
- https://issues.redhat.com/browse/KOGITO-5371

jomarko reviewed (approved):
I checked DMN and Scesim, some list of features I tried:
- zoom
- resize column
- DMN DRDs switching
- Scesim collection editor
- Shortcuts for DMN Data types, cut and paste DMN node, context menu
- Started Performance DMN test suite locally, seems fine

yesamer commented:
Hi @romartin! Code related to Scesim looks fine to me, thank you.
I started to test your `vsix` file, and it seems this issue https://issues.redhat.com/browse/KOGITO-4899 is back. (in this ticket: https://issues.redhat.com/browse/KOGITO-4953 you can find a video of the issue)

yesamer reviewed (commented):
Hi @romartin! Code related to Scesim looks fine to me, thank you.
I started to test your vsix file, and it seems this issue https://issues.redhat.com/browse/KOGITO-4899 is back. (in this ticket: https://issues.redhat.com/browse/KOGITO-4953 you can find a video of the issue)

romartin commented:
ok great, thanks for your feedback guys @jomarko @yesamer !!

So if I understand properly, DMN is fine, but it seems we're introducing a regression on SceSIm (see https://github.com/kiegroup/kogito-editors-java/pull/63#pullrequestreview-688360883). 

Soo I'm checking now the SceSim regression, also noticed an issue with the readonly mode in BPMN (maybe also applies to DMN), so I'm also looking at both issues now guys. 

If you find something else, please lemme know! 

@LuboTerifaj also waiting for your feedback on BPMN plz, before you leave on vacation, I owuld like to have QE approval here :)

Thanks again guys, appreciated!

romartin commented:
jenkins run fdb

domhanak commented:
The errors in FDB seems not to be related to these changes as #64 is also affected by same failures.

domhanak commented:
Hi @romartin , thanks for the PR I briefly checked BPMN editor in VSCode and all seems to be working well. The tests are passing. Keep in mind that final approval is left to @LuboTerifaj 

I went over these features:
- Checked that shortcuts from the list work
- Checked inline editing of node
- Checked Undo/Redo is working
- Checked save in general
- Checked zoom in/out
- Checked connectors handling
- Checked copy-paste with selections and nodes

karreiro reviewed (commented):
Thank you for the changes, @romartin.

The editor is much more stable. But, I've noticed that the Decision Services divider is losing the information when users resize the node or save/re-open the diagram:

![issue](https://user-images.githubusercontent.com/1079279/123105409-14f83500-d438-11eb-972e-a9de46f340ff.gif)

While it seems like a minor issue, it may impact the decision logic by changing a node from an encapsulated decision to an output decision.

Thanks again for the great work!

Josephblt commented:
Hello @romartin. I tested using the .Vsix provided. I could find only one minor issue regarding flow arrow drawing. If you compare with "old" versions you notice whenever the in/out nodes inside a lane are moved, the arrows are not drawn. But in your version they are and the results are somewhat strange. 

Here is a link with a video: https://drive.google.com/file/d/1TOpIjfVOXxE_4wnnvvJFW-ugdCGB2CFL/view

Also, I could not run the bpmn showcase from CLI.

Josephblt reviewed (changes requested):
Hello @romartin. I tested using the .Vsix provided. I could find only one minor issue regarding flow arrow drawing. If you compare with "old" versions you notice whenever the in/out nodes inside a lane are moved, the arrows are not drawn. But in your version they are and the results are somewhat strange.

Here is a link with a video: https://drive.google.com/file/d/1TOpIjfVOXxE_4wnnvvJFW-ugdCGB2CFL/view

Also, I could not run the bpmn showcase from CLI.

Other than that, nothing else to add. Great work. Thanks.

LuboTerifaj reviewed (commented):
Hello @romartin ,

I didn't find any issues in BPMN Process editor.

It seems all features work as expected.

One last thing that needs to be checked are the failing integration tests for standalone editors.

Thank you!

romartin commented:
ok **thanks all** guys!

Working on the issues you mentioned, hopefully no hard regressions found! :tada:  

Will let you konw once ready for review again.

Thansk!

romartin commented:
Hey @karreiro 

I have fixed the DMN issue when moving the decision service's line divider, good catch, thanks! Also fixed the readOnly mode for both editors, as it was not displaying properly the right canvas size before, but now everyting ok. I have updated the VSIX link in the PR description above, so please, can you give it a try again @karreiro ?

I am trying to get the approvals from each one of you one step at time, so for now, I think BPMN and DMN are ok to go, waiting for your feedback. Next step is looking at SceSim issue reported by @yesamer and the connectors issue reported by @Josephblt , will let you know guys once ready to test.

Thanks everyone for your collaboration!

romartin commented:
@yesamer Also fixed the regression you mentioned [KOGITO-4899](https://issues.redhat.com/browse/KOGITO-4899). Thanks for pointing it out!

So please, @karreiro @yesamer both DMN and SceSim ready to test again! I've updated the link for the VSIX in the description above: https://github.com/kiegroup/kogito-editors-java/pull/63#issue-669559318

Thanks!

romartin commented:
Hey @yesamer @LuboTerifaj @karreiro 

Friemdly reminder guys, this PR needs review again, and would be nice having all this huge codebase merged ASAP.

Thanks!

karreiro reviewed (approved):
Great stuff, @romartin. Thank you!

yesamer reviewed (approved):
@romartin Well done! LGTM!

romartin commented:
Hey @LuboTerifaj @Josephblt 

Just missing the review for the BPMN side from you guys, please. 

As I understand we have identitied 2 issues, not regressions, the one from lubo [1] and the one from wagner [2] which I will report once this code is in. Assming those 2 known issues, could you please give your feedback again? If no blocker issues, appreciate approving this PR, we need this merged urgently, it's blocking other stuff.

Thanks in advance!

[1] https://github.com/kiegroup/kogito-editors-java/pull/63#pullrequestreview-695201479
[2] https://github.com/kiegroup/kogito-editors-java/pull/63#pullrequestreview-693970586

domhanak commented:
@romartin please rerun the FDB, the VSCode tests are fixed now.

romartin commented:
Hey @LuboTerifaj 

Please remember to provide your feedback here. We've almost all approvals and green builds, so please, would appreciate early feedback.

romartin commented:
@LuboTerifaj sorry for bothering again, but please, a friendly reminder, this is really urgent! Waiting for your feedback, please consider this review. Thanks in advance!

LuboTerifaj reviewed (approved):
Hello @romartin ,

I didn't find any issues in BPMN editor.

Thank you for the great job! It looks good to me.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/test/java/org/kie/workbench/common/stunner/client/lienzo/shape/view/wires/WiresScalableContainerTest.java:
```diff
@@ -50,14 +47,14 @@
 
-    @Mock
-    private Attributes whAttributes;
+//    @Mock
```
done

romartin replied:
test case fixed :+1:

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/diagram/impl/DefaultDiagramEditor.java:
```diff
@@ -120,14 +120,8 @@ public void open(final Diagram item,
-                    height,
-                    new ViewCallback(callback));
+    public void open(Diagram item, int width, int height, DiagramViewerCallback<Diagram> callback) {
+        // TODO: lienzo-native
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-shapes/kie-wb-common-stunner-shapes-client/src/test/java/org/kie/workbench/common/stunner/shapes/client/view/PictureShapeViewTest.java:
```diff
@@ -61,6 +64,8 @@ public void checkPreDestructionRemovesResourcesFromDOM() {
+        verify(imageElement).remove();
     }
+
+ */
```
done

romartin replied:
I had been trying to make this lienzo based test case to work but it's failing due to internals about mocking elemental2 stuff... so I think it does not worth spending more time on mocking issues on this test case , it just tests a single condition anyway. Dropping it for now.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/diagram/impl/DefaultDiagramViewer.java:
```diff
@@ -147,6 +136,11 @@ protected void destroyInstances() {
 
+    @Override
+    public void open(Diagram item, int width, int height, DiagramViewerCallback<Diagram> callback) {
+        // TODO: lienzo-native
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionEditorImpl.java:
```diff
@@ -110,6 +110,11 @@ private S getSession() {
 
+    @Override
+    public void open(S item, int width, int height, SessionViewerCallback<Diagram> callback) {
+        // TODO lienzo-native
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionEditorImpl.java:
```diff
@@ -170,6 +175,11 @@ protected StunnerPreferencesRegistries getPreferencesRegistry() {
 
+        @Override
+        public void open(Diagram item, int width, int height, DiagramViewerCallback<Diagram> callback) {
+            // TODO lienzo-native
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionEditorPresenter.java:
```diff
@@ -180,4 +183,9 @@ private static void clearSelection(final EditorSession session) {
+
+    @Override
+    public void open(S item, int width, int height, SessionPresenterCallback<Diagram> callback) {
+        // TODO: lienzo-native
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionViewerImpl.java:
```diff
@@ -144,6 +149,11 @@ protected StunnerPreferencesRegistries getPreferencesRegistry() {
 
+        @Override
+        public void open(Diagram item, int width, int height, DiagramViewerCallback<Diagram> callback) {
+            // TODO lienzo-native
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionViewerPresenter.java:
```diff
@@ -121,4 +121,9 @@ protected void destroyToolbarInstace(final Toolbar<S> toolbar) {
+
+    @Override
+    public void open(S item, int width, int height, SessionPresenterCallback<Diagram> callback) {
+        // TODO lienzo-native
```
done

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionViewerImpl.java:
```diff
@@ -80,6 +80,11 @@ protected Diagram getDiagram() {
 
+    @Override
+    public void open(S item, int width, int height, SessionViewerCallback<Diagram> callback) {
+        // TODO lienzo-native
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/Context2D.java:
```diff
@@ -0,0 +1,423 @@
+    }
+
+    public ImageData createImageData(final ImageData data) {
+        // @FIXME this doesn't exist on Elm2
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/Context2D.java:
```diff
@@ -0,0 +1,423 @@
+        m_jso.setGlobalCompositeOperation((null != operation) ? operation.getValue() : null);
+    }
+
+//    public void setImageSmoothingEnabled(final boolean enabled)
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/NativeContext2D.java:
```diff
@@ -0,0 +1,681 @@
+            }
+        }
+
+        // @FIXME Add back in these pollyfills (or find better alterantive) (mdp)
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/NativeContext2D.java:
```diff
@@ -0,0 +1,681 @@
+
+    @JsOverlay
+    public final void setFillGradient(LinearGradientJSO grad) {
+        //createLinearGradient()
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/animation/AnimationScheduler.java:
```diff
@@ -0,0 +1,66 @@
+package com.ait.lienzo.client.core.animation;
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/config/LienzoCore.java:
```diff
@@ -0,0 +1,521 @@
+    }
+
+    private LienzoCore() {
+        // @FIXME need to figure out how to get the root DIV
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/config/LienzoCore.java:
```diff
@@ -0,0 +1,521 @@
+    }
+
+    public final void stack(final String message, final Throwable e) {
+        // @FIXME mdp
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/config/LienzoCore.java:
```diff
@@ -0,0 +1,521 @@
+    }
+
+    private final boolean examineNativeLineDashSupported() {
+        // @FIXME is this really neded now? Do we not have a minimum browser level now, that makes this redundant? (mdp)
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/animation/LayerRedrawManager.java:
```diff
@@ -0,0 +1,68 @@
+
+    private void kick(HTMLElement layerElement) {
+        if (!m_layers.isEmpty()) {
+            // @TODO check this works (mdp)
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/config/LienzoCorePlugin.java:
```diff
@@ -0,0 +1,164 @@
+
+    @Override
+    public final String getVersion() {
+        return "2.0.295-RELEASE";
```
class removed

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/event/AbstractNodeDragEvent.java:
```diff
@@ -0,0 +1,59 @@
+
+    private DragContext m_drag;
+
+//    public AbstractNodeDragEvent(final DragContext drag)
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/event/AbstractNodeDragEvent.java:
```diff
@@ -0,0 +1,59 @@
+        this.m_drag = m_drag;
+        setSource(sourceNode);
+
+//        m_x = MouseEventUtil.getRelativeX(event.clientX, getRelativeElement());
```
yes, done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/event/EventReceiver.java:
```diff
@@ -0,0 +1,9 @@
+package com.ait.lienzo.client.core.event;
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/event/AbstractNodeHumanInputEvent.java:
```diff
@@ -0,0 +1,148 @@
+package com.ait.lienzo.client.core.event;
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/event/OnEventHandlers.java:
```diff
@@ -0,0 +1,79 @@
+package com.ait.lienzo.client.core.event;
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/event/OnMouseEventHandler.java:
```diff
@@ -0,0 +1,10 @@
+package com.ait.lienzo.client.core.event;
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/i18n/MessageConstants.java:
```diff
@@ -0,0 +1,1811 @@
+ */
+public interface MessageConstants extends Constants {
+
+    // @FIXME no-op class for now, just to get things compiling (mdp)
```
done

romartin commented on lienzo-core/src/main/java/com/ait/lienzo/client/core/image/Image.java:
```diff
@@ -0,0 +1,382 @@
+package com.ait.lienzo.client.core.image;
```
done

romartin commented:
Hey @hasys - I've properly cleaned-up all your comments in thsi PR (most of them were out of date already). So hopefully I think it's everything ok now, if so, please apprecite your final approval for merging :) 

**Latest status** - I just merged again from master, and did again some smoke tests for all 3 editors (BPMN, DMN, SceSim) in vscode, everything looks ok. Just reported a couple of identified issues (not regressions), for further work:
- https://issues.redhat.com/browse/KOGITO-5556 : Toolbox menu has white colour but grey backround, when placed inside Lane CC @LuboTerifaj 
- https://issues.redhat.com/browse/KOGITO-5357 CC @Josephblt 

So guys, this looks ready to merge!! If no complains, I'll **proceed with merging** today 21th July.

Thanks everyone involved!

**REMINDER - DO NOT SQUASH!!**

hasys reviewed (approved):
Thank you @romartin, @handreyrc  fantastic job!

domhanak commented:
@kiegroup/gatekeepers I see all approvals are here, please merge and *DO NOT SQUASH* please. 


Thanks @romartin and everyone involved in the review. Great stuff!
